packages/protocol/thread/pause.ts

```diff
@@ -209,6 +205,17 @@ export class Pause {
   }
 
   addData(...datas: PauseData[]) {
+    const pauseId = this.pauseId!;
+    const point = this.point!;
+
+    // Event handlers must be called before processing the data.
+    // Calling _updateDataFronts() will add ValueFronts and mutate the underlying data.
+    if (pauseDataHandlers.length > 0) {
+      datas.forEach(data => {
+        pauseDataHandlers.forEach(handler => handler(pauseId, point, data));
+      });
+    }
```

Comment:
This (`addData`) is a better place to call pre-cache because more places may call this than just the `create` and `instantiate` methods.